accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+
+    *Note that both of the above cases can simultaneously apply because of
+    `dynamic` spreads.*
+
+*   If `element` is a `forElement` with `element` `p1` then:
+
+    Inference for the iterated expression and the controlling variable is done
+    as for the corresponding `for` or `await for` statement.
+
+    *   If the inferred set element type of `p1` is `S` then the inferred set
+        element type of `element` is `S`.
+
+    *   If the inferred key type of `p1` is `K` and the inferred key type of
+        `p1` is `V` then the inferred key and value types of `element` are `K`
+        and `V`.
+
+    *In other words, inference flows upwards from the body element. Note that
+    both of the above cases can validly apply because of `dynamic` spreads.*
+
+Finally, we define inference on a `setOrMapLiteral` `collection` as follows:
+
+*   If `collection` is a set literal, then the downwards context for inference
+    of the elements of `collection` is `Set<P>` where `P` may be `?` if
+    downwards inference does not constrain the type of `collection`.
+
+    *   If `P` is `?` then the static type of `collection` is `Set<T>` where `T`
+        is the upper bound of the inferred set element types of the elements.
+
+    *   Otherwise, the static type of `collection` is `Set<T>` where `T` is
+        determined by downwards inference.
+
+    *Note that the element inference will never produce a key/value type here
+    given that downwards context.*
+
+*   If `collection` is a map literal and downwards inference has statically
+    known type `Map<K, V>` then the downwards context for the elements of
+    `collection` is `Map<K, V>`.
+
+*   If `collection` is a map literal then the downwards context for the elements
+    of `collection` is `Map<Pk, Pv>` where `Pk` and `Pv` are determined by
+    downwards inference, and may be `?` if the downwards context does not
+    constrain one or both.
+
+    *   If `Pk` is `?` then the static key type of `collection` is `K` where `K`
+        is the upper bound of the inferred key types of the elements.
+
+    *   Otherwise the static key type of `collection` is `K` where `K` is
+        determined by downwards inference.
+
+    And:
+
+    *   If `Pv` is `?` then the static value type of `collection` is `V` where
+        `V` is the upper bound of the inferred value types of the elements.
+
+    *   Otherwise the static value type of `collection` is `V` where `V` is
+        determined by downwards inference.
+
+    *Note that the element inference will never produce a set element type here
+    given this downwards context.*
+
+*   If `collection` is not syntactically known to be a set or map literal, then
+    the downwards context for the elements of `collection` is `?`, and the disambiguation
+    is done as follows:
+
+    *   If all elements `ei` have a set element type `Ti`, and at least one
+        element does not have a key/value type pair, then `collection` is a set
```

Comment:
With `Null` generating no elements, change this to "and at least one element has a set element type and not a key/value type pair"

Reply:
Used "must be a set" and "must be a map" which I think covers this.